### ISSUE TITLE
Use dm-integrity with homed or some other method for data integrity

### ISSUE THREAD

**Issue by AidanGG** [systemd/systemd#24914]
### Component

systemd-homed

### Is your feature request related to a problem? Please describe

Trying to follow the blog post: https://0pointer.net/blog/authenticated-boot-and-disk-encryption-on-linux.html

My aim is to provide data authenticity for a user's home directory e.g. `/home/user` tied to their login credentials, as per the above blog post. Currently homed only sets up a LUKS dm-crypt volume which provides confidentiality tied to the login credentials through e.g. length-preserving modes like aes-xts.

Note that in the blog post, `/home` is left as an unencrypted dm-integrity volume, providing authenticity to the stored homed loop files tied to the system's keys. This is done to prevent double encryption of data within a home directory, by the system and user-derived keys. Adding authenticity checking to home directories would unfortunately result in two layers of integrity checking: one by the system's key at the `/home` level and another by the user-derived key at the `/home/user` level. If dm-integrity journalling is enabled at both levels this can result in a 4x write slowdown.

### Describe the solution you'd like

I can investigate adding a homectl `--luks-integrity` option to pass to cryptsetup's `--integrity` option to enable AEAD for home directories with a LUKS-managed key which unlocks a dm-crypt volume on top of a dm-integrity volume on top of the loop device. I imagine that something like

`--luks-cipher=chacha20 --luks-cipher-mode=random --luks-integrity=poly1305`

would achieve what we want. (Note that none the AEAD ciphers supported by the kernel have large nonces, so aes-xts-random with hmac-sha256 is probably the safest for now). Additionally, TRIM is not available in this set up.

### Describe alternatives you've considered

Perhaps checking integrity at the filesystem level would provide a sufficient level of confidence in data authenticity under encryption? For Btrfs within LUKS volumes for example, we would want to be able to configure the checksum used, so this could be another option:

`mkfs.btrfs --csum blake2 ...`

I remember there was some talk about adding authenticated hashing to Btrfs but I'm not sure that was ever merged.

### The systemd version you checked that didn't have the feature you are asking for

251

**Comment by poettering** [comment]
Adding dm-integrity support for homed's LUKS logic definitely makes sense.

So the question is what the integrity enforcement should achieve though. In my personal view of the world cryptographic integrity checks for /home/ are a way to ensure attackers cannot exploit the kernel fs drivers in some way and then acquire highest privs on the system. The kernel fs maintainers made very clear that they do not want to harden their fs implementation against rogue fs images, and it's a hard problem: the file systems have such complex data structures so that avoiding kernel exploits by manipulated file systems is very hard.

Thus I think it's important to establish integrity *below* the file system already, so that that trust is established first, and hence a vulnerable kernel fs implementation does not compromise security fully.

Now, HMAC-based dm-integrity can certainly guarantee that, but come at quite a performance cost. dm-integrity has a *much* smaller surface, so it'ss much less likely to be exploitable than a file system. btrfs' own checksum logic doesn#t really provide that at all however, the checksumming is on the fs layer, so you still have to trust the btrfs fs implementation to not be exploitable.

My current thinking is something else however: OPAL support in libcryptsetup, i.e. hw encryption. OPAL has a bad reputation, and I wouldn't trust it for confidentiality, but I think it should be fine for integrity. On typical SSDs OPAL comes basically at zero cost, it's built into all of them and typically comes at no performancy penalty.

So my hope is that we eventually can use a LUKS superblock with OPAL encryption on the lower level for /home/. And then on top we layer homed's LUKS stuff with software encryption.

TLDR: I don#t trust OPAl enough to keep my data secret, but I think it should be considered an OK choice for establishing trust on fs data structures.

But all this is just my personal take on the situation. I think adding direct dm-integrity support would be fine too. or adding support for btrfs checksumming is also totally OK.

**Comment by DemiMarie** [comment]
I don’t think that cryptographic integrity checks on home directories make sense unless one is using `dm-verity` (or at least `dm-integrity`) for `/`.

dm-integrity has two nasty problems:

1. It is vulnerable to replay attacks.
2. It is slow.

These can be somewhat worked around by dm-X, but that is still out of tree.  dm-X requires trusted nonvolatile storage to prevent replay attacks, which in practice is sadly unobtainable in the confidential computing model (it requires a secure element with high-endurance storage and low write latency), but might be possible in general.
